### PR TITLE
Box 3d eval

### PR DESCRIPTION
**Pull request by ncgaehle**
Evalulation for 3D Bounding Boxes

### CODE REVIEW

File: cityscapesscripts/evaluation/evalObjectDetection3D.py [hunk 1]
```diff
@@ -0,0 +1,1186 @@
...
+    logger.error(msg)
+    logger.info("========================")
+    logger.info("=== Stop evaluation ====")
+    logger.info("========================")
+
+    exit(1)
+
+class Box3DEvaluator:
+    """The Box3DEvaluator object contains the data as well as the parameters
+    for the evluation of the dataset.
```
Comment (nicolasj92) [inline cityscapesscripts/evaluation/evalObjectDetection3D.py]:
```suggestion
    for the evaluation of the dataset.
```

File: cityscapesscripts/evaluation/evalObjectDetection3D.py [hunk 2]
```diff
@@ -0,0 +1,1186 @@
...
+    :type results: dict
+    """
+
+    def __init__(
+        self,
+        evaluation_params: EvaluationParameters
+        ) -> None:
+
+        self.eval_params = evaluation_params
+        self._num_steps = 50
```
Comment (nicolasj92) [inline cityscapesscripts/evaluation/evalObjectDetection3D.py]: Why is this not in EvaluationParameters?

File: cityscapesscripts/evaluation/evalObjectDetection3D.py [hunk 3]
```diff
@@ -0,0 +1,1186 @@
...
+        logger.info("Evaluating images...")
+        # calculate stats for each image
+        self._calcImageStats()
+
+        logger.info("Calculate AP...")
+        # calculate 2D ap
+        self._calculateAp()
+
+        logger.info("Calculate TP stats...")
+        # calculate FP stats (center dist, size similarity, orientation score)
```
Comment (nicolasj92) [inline cityscapesscripts/evaluation/evalObjectDetection3D.py]:
```suggestion
        # calculate TP stats (center dist, size similarity, orientation score)
```

File: cityscapesscripts/evaluation/evalObjectDetection3D.py [hunk 4]
```diff
@@ -0,0 +1,1186 @@
...
+
+    def _calcImageStats(self) -> None:
+        """Internal method that calculates Precision and Recall values for whole dataset."""
+
+        # single threaded
+        results = []
+        for x in tqdm(self.gts.keys()):
+            results.append(self._worker(x))
+
+        # multi threaded
```
Comment (nicolasj92) [inline cityscapesscripts/evaluation/evalObjectDetection3D.py]: So multi thread does not work? Maybe we should delete the code then?

File: cityscapesscripts/evaluation/evalObjectDetection3D.py [hunk 5]
```diff
@@ -0,0 +1,1186 @@
...
+        # keep in mind that this will not work out of the box due the global interpreter lock
+        # with concurrent.futures.ThreadPoolExecutor(max_workers=2) as executor:
+        #    results = list(tqdm(executor.map(self._worker, self.gts.keys()), total=len(self.gts.keys())))
+
+        # update internal result dict with the curresponding results
```
Comment (nicolasj92) [inline cityscapesscripts/evaluation/evalObjectDetection3D.py]:
```suggestion
        # update internal result dict with the corresponding results
```

File: cityscapesscripts/evaluation/evalObjectDetection3D.py [hunk 6]
```diff
@@ -0,0 +1,1186 @@
...
+
+        self.results["eval_params"] = {
+            "labels": self.eval_params.labels_to_evaluate,
+            "min_iou_to_match": self.eval_params.min_iou_to_match,
+            "max_depth": self.eval_params.max_depth,
+            "step_size": self.eval_params.step_size,
+            "matching_method": modal_amodal_modifier
+        }
+
+        # calculate detection scores and them to results
```
Comment (nicolasj92) [inline cityscapesscripts/evaluation/evalObjectDetection3D.py]: add them?

File: cityscapesscripts/evaluation/plot3DResults.py [hunk 7]
```diff
@@ -0,0 +1,453 @@
...
+def create_table_row(
+        axis: Axes,
+        x_pos: float,
+        y_pos: float,
+        data_dict: dict,
+        title: str,
+        key: str,
+        subdict_key: str = None
+    ):
+    """Creates a row presentig scores for all classes in category ``key`` in ``data_dict``.
```
Comment (nicolasj92) [inline cityscapesscripts/evaluation/plot3DResults.py]:
```suggestion
    """Creates a row presenting scores for all classes in category ``key`` in ``data_dict``.
```

File: cityscapesscripts/evaluation/plot3DResults.py [hunk 8]
```diff
@@ -0,0 +1,453 @@
...
+    """Creates spider-chart with ``categories`` for all classes in ``accept_classes``.
+
+    Args:
+        axis (Axes): Axes-instances to use for the spider-chart
+        data_to_plot (dict): Dictionary containing ``categories`` as keys.
+        categories (list of str): List of category-names to use for the spider-chart.
+        accept_classes (list of str): List of class-names to use for the spider-chart.
+    """
+
+    # create lables
```
Comment (nicolasj92) [inline cityscapesscripts/evaluation/plot3DResults.py]:
```suggestion
    # create labels
```

**Review by nicolasj92 — APPROVED** [review]

File: cityscapesscripts/evaluation/evalObjectDetection3D.py (hunk 6 above)
Comment (ncgaehle) [reply cityscapesscripts/evaluation/evalObjectDetection3D.py]:
```suggestion
        # calculate detection scores and add them to results
```

File: cityscapesscripts/evaluation/evalObjectDetection3D.py (hunk 2 above)
Comment (ncgaehle) [reply cityscapesscripts/evaluation/evalObjectDetection3D.py]: moved to EvaluationParameters.

File: cityscapesscripts/evaluation/evalObjectDetection3D.py (hunk 4 above)
Comment (ncgaehle) [reply cityscapesscripts/evaluation/evalObjectDetection3D.py]: Removed multithreading.

**Review by nicolasj92 — APPROVED** [review]

**Pull request merged**